packages/api/src/API.ts

```diff
@@ -82,115 +85,146 @@ export class APIClass {
 	}
 
 	/**
 	 * Make a HEAD request
-	 * @param {string} apiName - The api name of the request
-	 * @param {string} path - The path of the request
-	 * @param {json} [init] - Request extra params
-	 * @return {Promise} - A promise that resolves to an object with response status and JSON data, if successful.
+	 * @param apiName - The api name of the request
+	 * @param path - The path of the request
+	 * @param [init] - Request extra params
+	 * @return A promise that resolves to an object with response status and JSON data, if successful.
 	 */
-	head(apiName, path, init): Promise<any> {
+	head(
+		apiName: string,
+		path: string,
+		init: { [key: string]: any }
+	): Promise<any> {
 		return this._restApi.head(apiName, path, init);
 	}
 
 	/**
 	 * Checks to see if an error thrown is from an api request cancellation
-	 * @param {any} error - Any error
-	 * @return {boolean} - A boolean indicating if the error was from an api request cancellation
+	 * @param error - Any error
+	 * @return If the error was from an api request cancellation
 	 */
-	isCancel(error) {
+	isCancel(error: any): boolean {
 		return this._restApi.isCancel(error);
 	}
 	/**
 	 * Cancels an inflight request
-	 * @param {any} request - request to cancel
-	 * @return {boolean} - A boolean indicating if the request was cancelled
+	 * @param request - request to cancel
+	 * @param [message] - custom error message
+	 * @return If the request was cancelled
 	 */
-	cancel(request: Promise<any>, message?: string) {
+	cancel(request: Promise<any>, message?: string): boolean {
 		return this._restApi.cancel(request, message);
 	}
 
 	/**
 	 * Getting endpoint for API
-	 * @param {string} apiName - The name of the api
-	 * @return {string} - The endpoint of the api
+	 * @param apiName - The name of the api
+	 * @return The endpoint of the api
 	 */
-	async endpoint(apiName) {
+	async endpoint(apiName: string): Promise<string> {
 		return this._restApi.endpoint(apiName);
 	}
 
 	/**
 	 * to get the operation type
 	 * @param operation
 	 */
-	getGraphqlOperationType(operation) {
+	getGraphqlOperationType(operation: GraphQLOperation): OperationTypeNode {
 		return this._graphqlApi.getGraphqlOperationType(operation);
 	}
 
 	/**
 	 * Executes a GraphQL operation
 	 *
-	 * @param {GraphQLOptions} GraphQL Options
-	 * @param {object} additionalHeaders headers to merge in after any `graphql_headers` set in the config
-	 * @returns {Promise<GraphQLResult> | Observable<object>}
+	 * @param options - GraphQL Options
+	 * @param [additionalHeaders] - headers to merge in after any `graphql_headers` set in the config
+	 * @returns An Observable if queryType is 'subscription', else a promise of the graphql result from the query.
 	 */
-	graphql(
+	graphql<T = any>(
 		options: GraphQLOptions,
 		additionalHeaders?: { [key: string]: string }
-	): Promise<GraphQLResult> | Observable<object> {
+	): T extends GraphQLQuery<T>
+		? Promise<GraphQLResult<T>>
+		: Observable<GraphQLResult<T>>;
```

Comment:
I think we should default the return type to the same one we have now for backward compatibility when type when `T` is neither `GraphQLQuery<T>` nor `GraphQLSubscription<T>`. 

```suggestion
	): T extends GraphQLQuery<T>
		? Promise<GraphQLResult<T>>
		: T extends GraphQLSubscription <T>
    ? Observable<GraphQLResult<T>>
    : Promise<GraphQLResult> | Observable<object>;
```